### PR TITLE
StaticReflectionClass PHP7.2 issue

### PR DESCRIPTION
**Pull request by alexpott**
Unfortunately PHP7.2 \ReflectionClass has slightly different semantics which means that \Doctrine\Common\Reflection\StaticReflectionClass has issues working on both PHP7.1 (and earlier) and PHP7.2.

The error you get on PHP7.2 is:
`Declaration of Doctrine\Common\Reflection\StaticReflectionClass::newInstance($args) should be compatible with ReflectionClass::newInstance(...$args)`

An earlier PR on this problem was https://github.com/doctrine/common/pull/775#pullrequestreview-21471640

### CODE REVIEW

**Comment by alexpott** [comment]
The first commit is a tests only situation where we should see PHP7.1 pass but PHP7.2 fail.

**Comment by Majkl578** [comment]
This looks like a PHP 7.2 BC break, have you reported it? https://3v4l.org/7igHL

**Comment by alexpott** [comment]
@Majkl578 yep - it was reported in https://github.com/php/php-src/pull/2363#issuecomment-279246184 - I guess the comment there suggests that the rollback should have gone to master too.

**Comment by Ocramius** [comment]
Relevant: https://github.com/Roave/BetterReflection/issues/317

File: lib/Doctrine/Common/Reflection/StaticReflectionClass.php [hunk 1]
```diff
@@ -22,412 +22,417 @@
...
-    /**
-     * {@inheritDoc}
-     */
-    public function __toString()
-    {
-        throw new ReflectionException('Method not implemented');
+if (version_compare(phpversion(), '7.2', '<')) {
+    include __DIR__ . '/legacy/StaticReflectionClass.php';
+}
+else {
```
Comment (EclipseGc) [inline lib/Doctrine/Common/Reflection/StaticReflectionClass.php]: Let's do the version check in a trait definition and just use that trait in this class to swap out the method.

**Comment by alexpott** [comment]
Thanks @Ocramius that's super simple and appears to work for both 7.1 and 7.2. Yay!

**Comment by dawehner** [comment]
@Ocramius You are a genius. Thank you!

**Comment by Majkl578** [comment]
I'd rather wait for PHP to get this fixed on their side... There's still time to fix it before 7.2 goes stable.
There's already a bug report for this: https://bugs.php.net/bug.php?id=74292

**Comment by EclipseGc** [comment]
But this won't work for 5.5 will it?

**Comment by Ocramius** [comment]
5.5 is not supported anymore?

**Comment by EclipseGc** [comment]
Right, I was making the point for Drupal which is still supporting 5.5 but would also like to support 7.2. The trait based approach (with an include for the 7.2 specific code) is the best solution for our needs I think.

**Comment by alexpott** [comment]
So PHP have decided to rollback the BC break.

**Pull request closed without merging**